docs/replication.md

```diff
@@ -0,0 +1,145 @@
+- Couchdb `_changes` and by extension replication can be either by polling `?since=xx&limit=xx` can return `[]`, or continuous (SSE / COMET)
+- In couchdb benchmarking, we understood that the number of couch databases only use a bit of disk space but no RAM or CPU **as long as the database is not used**. Having a continuous replication active force us to keep the database file open and will starve RAM & FD usage. Moreover, continuous replication costs a lot (cf. ScienceTeam benchmark). To permit an unlimited number of inactive user on a single cozy-stack process, **the stack should avoid continuous replication from couchdb**.
+- Two way replication is simply two one way replications
+
+
+## Routes used by replication
+
+To be a source of replication, the stack only need to support the following route:
+
+- `GET :source/:docid?open_revs=xxxx` get revisions of a document
+- `GET  :source/_changes` get a list of ID -> New Revision since a given sequence number
+
+To be a target of replication, the stack need to support the following routes:
+
+- `POST :target/_revs_diff` takes a list of ID -> Rev and returns which one are missing.
+- `POST :target/_bulk_docs` create several documents at once
+- `POST :target/_ensure_full_commit` ensure the documents are written to disk. This is useless if couchdb is configured without delayed write (default), but remote couchdb will call, so the stack should return the expected 201
+
+In both case, we need to support
+
+- `PUT    :both/_local/:revdocid` to store the current sequence number.
+
+# Stack Sync API exploration
+
+
+## Easy part: 1 db/doctype on stack AND remote, no binaries
+We *just* need to implement the routes described above by proxying to the underlying couchdb database.
+
+```javascript
+var db = new PouchDB("my-local-contacts")
+db.replicate.from('https://bob.cozycloud.cc/data/contacts')
+db.replicate.to('https://bob.cozycloud.cc/data/contacts')
+```
+
+To suport this we need to :
+- Proxy `_changes` route with since, limit, feed=normal. Refuse all filter parameters. [(Doc)](http://docs.couchdb.org/en/2.0.0/api/database/changes.html)
+- Add support of `open_revs` query parameter to `GET /data/:doctype/:docid` [(Doc) ](http://docs.couchdb.org/en/2.0.0/api/document/common.html?highlight=open_revs#get--db-docid)
+- Proxy the `_revs_diff` [(Doc)](http://docs.couchdb.org/en/2.0.0/api/database/misc.html#db-revs-diff) and `_bulk_docs` [(Doc)](http://docs.couchdb.org/en/2.0.0/api/database/bulk-api.html) routes
+- Have `_ensure_full_commit` [(Doc)](http://docs.couchdb.org/en/2.0.0/api/database/compact.html#db-ensure-full-commit) returns 201
+
+This will cover documents part of the Devices use case.
+
+With a way to list doctype, this could also cover Backup and Moving.
+
+
+## Continuous replication
+
+It is impossible to implement it by simply proxying to couchdb (see unlimited inactive users). But neither the Device nor the Backup option need open continuous replication.
```

Comment:
Cozy-desktop uses continuous replication (well, only a continuous changes-feed), so I disagree here to say that Device doesn't need open continuous replication.